File: 2369. Check if There is a Valid Partition For The Array.py
from typing import List
class Solution:
    def validPartition(self, nums: List[int]) -> bool:
        dp = [True, False, nums[0] == nums[1]]
        for i in range(2, len(nums)):
            cur = False
            if nums[i] == nums[i - 1] and dp[1]:
                cur = True
            elif nums[i] == nums[i - 1] == nums[i - 2] and dp[0]:
                cur = True
            elif nums[i] - nums[i - 1] == nums[i - 1] - nums[i - 2] == 1 and dp[0]:
                cur = True
            dp[0], dp[1], dp[2] = dp[1], dp[2], cur
        return dp[2]

File: 2369. Check if There is a Valid Partition For The Array.kt
class Solution {
    fun validPartition(nums: IntArray): Boolean {
        val dp = booleanArrayOf(true, false, nums[0] == nums[1])
        for (i in 2 until nums.size) {
            val cur = if (nums[i] == nums[i - 1] && dp[1]) {
                true
            } else if (nums[i] == nums[i - 1] && nums[i - 1] == nums[i - 2] && dp[0]) {
                true
            } else if (nums[i] - nums[i - 1] == 1 && nums[i - 1] - nums[i - 2] == 1 && dp[0]) {
                true
            } else {
                false
            }
            dp[0] = dp[1]
            dp[1] = dp[2]
            dp[2] = cur
        }
        return dp[2]
    }
}